library/core/src/ptr/mod.rs

```diff
@@ -1744,54 +1749,65 @@ pub const unsafe fn write_unaligned<T>(dst: *mut T, src: T) {
-/// leaves the memory in `src` unchanged.
-///
-/// Volatile operations are intended to act on I/O memory, and are guaranteed
-/// to not be elided or reordered by the compiler across other volatile
-/// operations.
-///
-/// # Notes
-///
-/// Rust does not currently have a rigorously and formally defined memory model,
-/// so the precise semantics of what "volatile" means here is subject to change
-/// over time. That being said, the semantics will almost always end up pretty
-/// similar to [C11's definition of volatile][c11].
-///
-/// The compiler shouldn't change the relative order or number of volatile
-/// memory operations. However, volatile memory operations on zero-sized types
-/// (e.g., if a zero-sized type is passed to `read_volatile`) are noops
-/// and may be ignored.
+/// Performs a volatile read of the value from `src` without moving it.
+///
+/// Rust does not currently have a rigorously and formally defined memory model, so the precise
+/// semantics of what "volatile" means here is subject to change over time. That being said, the
+/// semantics will almost always end up pretty similar to [C11's definition of volatile][c11].
+///
+/// Volatile operations are intended to act on I/O memory. As such, they are considered externally
+/// observable events (just like syscalls), and are guaranteed to not be elided or reordered by the
+/// compiler across other externally observable events. With this in mind, there are two cases of
+/// usage that need to be distinguished:
+///
+/// - When a volatile operation is used for memory inside an [allocation], it behaves exactly like
+///   [`read`], except for the additional guarantee that it won't be elided or reordered (see
+///   above). This implies that the operation will actually access memory and not e.g. be lowered to
+///   a register access or stack pop. Other than that, all the usual rules for memory accesses
+///   apply. In particular, just like in C, whether an operation is volatile has no bearing
+///   whatsoever on questions involving concurrent access from multiple threads. Volatile accesses
+///   behave exactly like non-atomic accesses in that regard.
+///
+/// - Volatile operations, however, may also be used access memory that is _outside_ of any Rust
+///   allocation. The main use-case is CPU and peripheral registers that must be accessed via an I/O
+///   memory mapping, most commonly at fixed addresses reserved by the hardware. These often have
+///   special semantics associated to their manipulation, and cannot be used as general purpose
+///   memory. Here, any address value is possible, including 0 and [`usize::MAX`], so long as the
+///   semantics of such a read are well-defined by the target hardware. The access must not trap. It
+///   can (and usually will) cause side-effects, but those must not affect Rust-allocated memory in
+///   any way. In this use-case, the pointer validity rules in the [module documentation][mod-docs]
+///   may not apply.
```

Comment:
It seems odd to link to a different place here than where `[valid]` links below, and to use a different term.

```suggestion
///   any way. In this use-case, the pointer does *not* have to be [valid] for reads.
```